crates/uv-resolver/src/flat_index.rs

```diff
@@ -155,11 +171,16 @@ impl FlatIndex {
         tags: &Tags,
         hasher: &HashStrategy,
         no_binary: &NoBinary,
+        no_build: &NoBuild,
     ) -> WheelCompatibility {
         // Check if binaries are allowed for this package.
         let no_binary = match no_binary {
             NoBinary::None => false,
-            NoBinary::All => true,
+            NoBinary::All => match no_build {
+                // Allow `all` to be overridden by specific build exclusions
+                NoBuild::Packages(packages) => !packages.contains(&filename.name),
+                _ => true,
+            },
```

Comment:
Could we instead create a struct that couples `NoBinary` and `NoBuild`, and lets us query for the `no_binary` or `no_build` status, abstracting away this repeated logic? It seems really easy to miss a site as-is.

Reply:
Yeah we should totally do that as noted at https://github.com/astral-sh/uv/pull/4067#issuecomment-2150906635 but I'd rather do it in a follow-up.

I didn't want to invest in a refactor until we were on the same page about the behavior. I could do it here too if you think this makes sense.

Reply:
I think the behavior seems ok. Do you want to do it in this PR or na?

Reply:
I don't really mind doing it here or at least writing the follow-up before merge.

Reply:
See #4284 — opted for a separate pull request for review purposes.